### ISSUE TITLE
Code review 12.6.2018

### ISSUE THREAD

**Issue by petyrek** [strvcom/dep-manager-web#12]
YO YO YO, i did a little review, hope you will find something useful here

## SVGs
not minimized
Dump the svg code in here https://jakearchibald.github.io/svgomg/ and it will spit out minimized output

## CSS reset
This could be installed as dependency and simply included, or it is a part of the `polished` package.
Though there's not necessarily anything with what you're doing. But maybe that people can mess with it this way, when it's not encapsulated, but they really shouldn't.

## Flow
Lots of type errors everywhere, broken imports.
To fix imports do something like
`/flow-typed/ignore.js`
with the following contents
```
// @flow

declare module "react-swipeable-views" {
  declare module.exports: any
}
```

Inference doesn't work here
`export default class Dashboard extends Component<*, *> {`
as it has too little information.
Full shape of state cannot be infered from
```
state = {
    projects: {},
    latestLibraries: {},
    projectLibraryRelation: [],
    filteredProjects: [],
    filteredLibraries: [],
    isLoading: true
  };
``` 
so Ideally you want to be a little more explicit to ensure type safety.

## Apollo
What you have in `/src/App.js` using the `<Query>` component seems to me like a horrible pattern. Unclear what happens and seems like it could be buggy when this component rerenders, because you are invoking side effects from within the render method. I would be ultra carefull with code like that.

Another "misuse" example can be found at `containers/Dashboard/index.js`
The way you are handling the requests is not the graphql way. You should take advantage of the `graphql` HOC and you can then get rid of code like fetching on `componentDidMount`.

## Axios
no reason for this, drop in favor of fetch, which is a standard API.

## File size
Take `containers/Dashboard/index.js` this file is way too big.
You are mixing styled compoents with React component.
Create a `styled.js` file and put all the styled components there, so they are not in the way.

There are also route definitions, some of the state is missing in type annotations and it is hard to follow.
This needs to be split into multiple files.

## Imports
instead of this
`import LibraryDetail from "../../containers/LibraryDetail";`
add `.env` file with `NODE_PATH=src/` it will allow you to import it like this `"containers/LibraryDetail"`

## Routing
Ideally you would have `/pages` folder that would contain all the top level route components. In each of these folders you would have components/containers specific to this page. The shared ones would live in `/components` or `/containers` this is extremely important for project to be able to grow and still remain maintainable.

Also I would suggest to create `/pages/urls.js` file where you would define all the routes. That way you can get rid of magic constant urls and a potential to break links.

## More on structure
What you have in `/data` is okay, but i would suggest to break it down per entity.

So you would end up with
```
/data
  /Animal
    /apis.js
    /groups.js
    /helpers.js
    /queries.js
  /Books
    /apis.js
    /groups.js
    /helpers.js
    /queries.js
```
this again, ensures your files are not too big and easy to follow.

# What you can do with apollo, some inspiration from awayco, because i have been playing with GQL for quite a while there.
`src/types/graphql.js`
what a Query looks like, Data is type of data returned from the HOC, if the query takes params, they are the second type argument
```
export type Query<Data, Params = empty> = {|
  gql: any,
  selector?: string[],
  transform?: any => Data,
  pagination?: string[],
  options?: Function,
|}
```

`src/data/reservations/graphql/queries.js`
Here is then a concrete instance of the Query type
```
import gql from "graphql-tag"
import type { Query } from "types/graphql"

export const reservationCountQuery: Query<number> = {
  gql: gql`
    query reservationCountQuery {
      reservation {
        list(filter: { limit: 9999999, status: confirmed }) {
          data {
            id
          }
        }
      }
    }
  `,
  transform: length,
  selector: ["reservation", "list", "data"],
}
```

I extended the behavior of the `graphql` HOC, so this code is quite complicated, but it allows for very easy use, ill show below.
```
// @flow
// import * as React from "react"
import { graphql } from "react-apollo"
import type { Query } from "types/graphql"
import { compose, branch, renderComponent, type HOC } from "recompose"
import { getRouterVariables } from "common/graphql/routing"
import { transformResponse } from "common/graphql/helpers"
import withLoader from "hoc/withLoader"
import withEmpty from "hoc/withEmpty"
import NotFoundPage from "pages/NotFound"
import { pathSatisfies, pathOr, endsWith, find, always } from "ramda"
import { getPageInfo, getPaginationVariables, mergeData } from "./pagination"

const notFound = compose(
  find(pathSatisfies(endsWith("NOT_FOUND"), ["data", "code"])),
  pathOr([], ["error", "graphQLErrors"]),
)

// eslint-disable-next-line
type Config<Outter, Variables> = {
  noLoader?: boolean,
  noEmpty?: boolean,
  variables?: Outter => Variables,
  config?: {
    skip: Outter => boolean,
  },
}

type Added<Data> = {
  data: Data,
  loading: boolean,
  loadingMore: boolean,
  count: number,
  loadMore: void => void,
  notFound: boolean,
}

function withQuery<Outter, Data, Variables>(
  query: Query<Data, Variables>,
  configuration: Config<Outter, Variables> = {},
  // $ExpectError
): HOC<{ ...$Exact<Added<Data>>, ...$Exact<Outter> }, Outter> {
  const { gql, selector, transform, pagination, options = always({}) } = query
  const {
    noLoader = false,
    noEmpty = false,
    variables = always({}),
    config = {},
  } = configuration

  const customGraphqlHoc = graphql(gql, {
    props: props => ({
      data: transformResponse(selector, transform)(props),
      // loading | setVariables | refetch
      loading:
        props.data.networkStatus === 1 ||
        props.data.networkStatus === 2 ||
        props.data.networkStatus === 4,
      loadingMore: props.data.networkStatus === 3,
      notFound: notFound(props.data),
      count: pagination && getPageInfo(pagination, props.data).totalCount,
      loadMore: () =>
        pagination &&
        !props.data.loading &&
        getPageInfo(pagination, props.data).hasNextPage &&
        props.data.fetchMore({
          variables: getPaginationVariables(pagination, props.data),
          updateQuery: (prev, { fetchMoreResult }) =>
            mergeData(prev, fetchMoreResult, pagination),
        }),
    }),
    options: props => ({
      notifyOnNetworkStatusChange: true, // to update loading when fetching more
      variables: {
        ...getRouterVariables(props),
        ...getPaginationVariables(pagination),
        ...variables(props),
      },
      ...options(props),
    }),
    ...config,
  })

  return compose(
    customGraphqlHoc,
    ...(noLoader ? [] : [withLoader(props => props.loading)]),
    branch(x => x.notFound, renderComponent(NotFoundPage)),
    ...(noEmpty ? [] : [withEmpty()]),
  )
}

export default withQuery
```

And finally, how very easily it can be used
```
// @flow
import React from 'react'
import type { HOC } from 'recompose'
import { NavLink } from "componentsStyled/Typography"
import withQuery from "hoc/withQuery"
import { isNil } from "ramda"
import { reservationCountQuery } from "data/reservations/graphql/queries"
import urls from "pages/urls"

const ReservationsLink = ({ data }) => (
  <NavLink to={urls.reservations}>
    reservations&nbsp;{isNil(data) ? "" : `(${data})`}
  </NavLink>
)

const enhancer:HOC<*, {}> = withQuery(reservationCountQuery, {
  noEmpty: true,
  noLoader: true,
})

export default enhancer(ReservationsLink)
```

I defined something similar for mutations, which then allows for a seamless refetching of queries and so on.

**Comment by bsunderhus** [comment]
Ok let me add a list here of what has been done and is still missing

- [x] SVG
- [x] CSS reset (5cef36c15ffcaf63e2da3411ef47110b3a2f7c3a)
- [x] Flow
- [x] Apollo
- [x] Axios
- [x] File size
- [x] Routing
- [x] More on structure

## SVGs (be77a6d381b81d32dea20429d0e51f78e77fda85)

The [create-react-app](https://facebook.github.io/create-react-app/) handles this by using [svgr](https://github.com/smooth-code/svgr) under the hood, transforming the SVG into a React component and also applying all optimizations on it.

## Flow

The whole application has been converted to typescript from this commit 48a5d11e770ff739a4c258e81b9479be1a5e6902 forward. Most of the problems with the lack of good typings are still being fixed.

## Apollo

The whole Apollo structure used in this project has been reviewed and now the [Apollo Client](https://www.apollographql.com/docs/react/) serves as a single source of truth for the whole project.

This has been quite a challenge to the project since it not only analyses the GitHub GraphQL API but also analyses the NPM API through [npms](https://api-docs.npms.io).

## Axios

This has been completely removed!

## File Size

Most of the huge files were already broken into small ones, such as `Dashboard`

## Routing

The whole routing of the application has been remodeled to take bigger advantages of redirections, switches, and conditional routing.
